src/optimize_1q_gates.rs

```diff
@@ -0,0 +1,102 @@
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+const PI: f64 = std::f64::consts::PI;
+
+//     Return a triple theta, phi, lambda for the product.
+//
+//         u3(theta, phi, lambda)
+//            = u3(theta1, phi1, lambda1).u3(theta2, phi2, lambda2)
+//            = Rz(phi1).Ry(theta1).Rz(lambda1+phi2).Ry(theta2).Rz(lambda2)
+//            = Rz(phi1).Rz(phi').Ry(theta').Rz(lambda').Rz(lambda2)
+//            = u3(theta', phi1 + phi', lambda2 + lambda')
+//
+//         Return theta, phi, lambda.
+#[pyfunction]
+pub fn compose_u3_rust(
+    theta1: f64,
+    phi1: f64,
+    lambda1: f64,
+    theta2: f64,
+    phi2: f64,
+    lambda2: f64,
+) -> [f64; 3] {
+    let q = [(theta1 / 2.0).cos(), 0., (theta1 / 2.0).sin(), 0.];
+    let r = [
+        ((lambda1 + phi2) / 2.0).cos(),
+        0.,
+        0.,
+        ((lambda1 + phi2) / 2.0).sin(),
+    ];
+    let s = [(theta2 / 2.0).cos(), 0., (theta2 / 2.0).sin(), 0.];
+
+    // Compute YZY decomp (q.r.s in variable names)
+    let temp: [f64; 4] = [
+        r[0] * q[0] - r[1] * q[1] - r[2] * q[2] - r[3] * q[3],
+        r[0] * q[1] + r[1] * q[0] - r[2] * q[3] + r[3] * q[2],
+        r[0] * q[2] + r[1] * q[3] + r[2] * q[0] - r[3] * q[1],
+        r[0] * q[3] - r[1] * q[2] + r[2] * q[1] + r[3] * q[0],
+    ];
+
+    let out: [f64; 4] = [
+        s[0] * temp[0] - s[1] * temp[1] - s[2] * temp[2] - s[3] * temp[3],
+        s[0] * temp[1] + s[1] * temp[0] - s[2] * temp[3] + s[3] * temp[2],
+        s[0] * temp[2] + s[1] * temp[3] + s[2] * temp[0] - s[3] * temp[1],
+        s[0] * temp[3] - s[1] * temp[2] + s[2] * temp[1] + s[3] * temp[0],
+    ];
+
+    // out is now in YZY decomp, make into ZYZ
+    let mat: [f64; 9] = [
+        1. - 2. * out[2] * out[2] - 2. * out[3] * out[3],
+        2. * out[1] * out[2] - 2. * out[3] * out[0],
+        2. * out[1] * out[3] + 2. * out[2] * out[0],
+        2. * out[1] * out[2] + 2. * out[3] * out[0],
+        1. - 2. * out[1] * out[1] - 2. * out[3] * out[3],
+        2. * out[2] * out[3] - 2. * out[1] * out[0],
+        2. * out[1] * out[3] - 2. * out[2] * out[0],
+        2. * out[2] * out[3] + 2. * out[1] * out[0],
+        1. - 2. * out[1] * out[1] - 2. * out[2] * out[2],
+    ];
+
+    // Grab the euler angles
+    let mut euler: [f64; 3] = [0., 0., 0.];
+    if mat[8] < 1.0 {
+        if mat[8] > -1.0 {
+            euler[0] = mat[5].atan2(mat[2]);
+            euler[1] = (mat[8]).acos();
+            euler[2] = mat[7].atan2(-mat[6]);
+        } else {
+            euler[0] = -1. * (mat[3].atan2(mat[4]));
+            euler[1] = PI;
+        }
+    } else {
+        euler[0] = mat[3].atan2(mat[4]);
+    }
+    for k in &mut euler {
+        if k.abs() < 1e-15 {
+            *k = 0.0;
+        }
+    }
+
+    let out_angles: [f64; 3] = [euler[1], phi1 + euler[0], lambda2 + euler[2]];
+    out_angles
```

Comment:
Thanks for this tip and the others above too. I'm not sure how to do this part properly (still learning the basics). Maybe something like this?
```
#[pyfunction]
pub fn compose_u3_rust(
    py: Python,
    theta1: f64,
    . . .
    ) -> &PyArray1<f64> {
    . . .
    . . .
    vec![euler[1], phi1 + euler[0], lambda2 + euler[2]].into_pyarray(py)
```

And then calling `tuple(compose_u3_rust(...))` in Python so that `compose_u3` returns a tuple like it did originally?

I may just leave the code in the PR unchanged for now, but would be interested if you have an example handy of the right way to no-copy return. Thanks again.

Reply:
Yeah, in this case I don't think it'll matter since it's a 3 element array so leaving it as is probably fine. Although it might be interesting to benchmark just to see if it's noticeable especially as this kind of optimization can be deceptive some times and the copy to a tuple might actually end up being faster (it would definitely faster to use numpy for a large array which is all I was trying to say here).

But yeah your suggestion is probably the easiest way (although I've found the return type handling for the ffi to be finicky sometimes and often convert to a `PyObject` for the return, see https://github.com/Qiskit/qiskit-terra/blob/main/src/results/marginalization.rs#L180). For rust-numpy to be able to pass the pointer to numpy the array needs to be heap allocated, so a `Vec` is the most common type to use for this. The other thing you could do is use a [boxed](https://doc.rust-lang.org/std/boxed/index.html) array to keep it fixed size, the rust numpy binding supports creating a numpy array directly from a boxed array: https://docs.rs/numpy/latest/numpy/convert/trait.IntoPyArray.html#impl-IntoPyArray-for-Box%3C%5BT%5D%3E as for the return type to really get the advantage I think we'd have to change `compose_u3` to return an ndarray since doing `tuple(compose_u3_rust())` would be doing essentially the same thing just via numpy instead of the rust extension. 

The other thing you could do is create a custom python class in rust that wraps the array and implements the sequence protocol on top of it. Then you only have the copy/conversion overhead on `__getitem__` (to go from a f64->pyfloat) and avoid the tuple creation time. But this also is a tradeoff because depending on how things are accessed you might end up creating more overhead if you access elements more than once. https://github.com/Qiskit/qiskit-terra/pull/7977/files#diff-e33d89bd9a7d958f9784e026a1f585768a6e517dabf8d519391800609c8a4a34 is an example of this from #7977